devito/tools/memoization.py

```diff
@@ -1,135 +1,217 @@
-from collections.abc import Callable, Hashable
-from functools import lru_cache, partial
-from itertools import tee
-from typing import TypeVar
+from collections.abc import Hashable, Iterator
+from functools import lru_cache, partial, update_wrapper
+from threading import RLock, local
+from typing import Callable, Concatenate, Generic, ParamSpec, TypeVar
 
-__all__ = ['memoized_func', 'memoized_meth', 'memoized_generator', 'CacheInstances']
 
+__all__ = ['memoized_meth', 'memoized_generator', 'CacheInstances']
 
-class memoized_func:
-    """
-    Decorator. Caches a function's return value each time it is called.
-    If called later with the same arguments, the cached value is returned
-    (not reevaluated). This decorator may also be used on class methods,
-    but it will cache at the class level; to cache at the instance level,
-    use ``memoized_meth``.
 
-    Adapted from: ::
+# Type variables for memoized method decorators
+InstanceType = TypeVar('InstanceType', contravariant=True)
+ParamsType = ParamSpec('ParamsType')
+ReturnType = TypeVar('ReturnType', covariant=True)
 
-        https://wiki.python.org/moin/PythonDecoratorLibrary#Memoize
-    """
 
-    def __init__(self, func):
-        self.func = func
-        self.cache = {}
-
-    def __call__(self, *args, **kw):
-        if not isinstance(args, Hashable):
-            # Uncacheable, a list, for instance.
-            # Better to not cache than blow up.
-            return self.func(*args, **kw)
-        key = (self.func, args, frozenset(kw.items()))
-        if key in self.cache:
-            return self.cache[key]
-        else:
-            value = self.func(*args, **kw)
-            self.cache[key] = value
-            return value
-
-    def __repr__(self):
-        """Return the function's docstring."""
-        return self.func.__doc__
-
-    def __get__(self, obj, objtype):
-        """Support instance methods."""
-        return partial(self.__call__, obj)
-
-
-class memoized_meth:
+class memoized_meth(Generic[InstanceType, ParamsType, ReturnType]):
+    """
+    Decorator for a cached instance method. There is one cache per thread stored
+    on the object instance itself.
     """
-    Decorator. Cache the return value of a class method.
-
-    Unlike ``memoized_func``, the return value of a given method invocation
-    will be cached on the instance whose method was invoked. All arguments
-    passed to a method decorated with memoize must be hashable.
-
-    If a memoized method is invoked directly on its class the result will not
-    be cached. Instead the method will be invoked like a static method: ::
 
-        class Obj:
-            @memoize
-            def add_to(self, arg):
-                return self + arg
-        Obj.add_to(1) # not enough arguments
-        Obj.add_to(1, 2) # returns 3, result is not cached
+    def __init__(self, meth: Callable[Concatenate[InstanceType, ParamsType],
+                                      ReturnType]) -> None:
```

Comment:
```suggestion
    def __init__(
        self,
        meth: Callable[Concatenate[InstanceType, ParamsType], ReturnType]
    ) -> None:
```

Reply:
This would be my preferred style for all function signatures that are too long due to type hinting